network/ip_tracker.go

```diff
@@ -0,0 +1,360 @@
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	delete(i.connected, nodeID)
+	i.removeGossipableIP(nodeID)
+}
+
+func (i *ipTracker) OnValidatorAdded(nodeID ids.NodeID, _ *bls.PublicKey, _ ids.ID, _ uint64) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	i.onValidatorAdded(nodeID)
+}
+
+func (i *ipTracker) onValidatorAdded(nodeID ids.NodeID) {
+	if i.manuallyTracked.Contains(nodeID) {
+		return
+	}
+
+	i.validators.Add(nodeID)
+	ip, connected := i.connected[nodeID]
+	if !connected {
+		return
+	}
+
+	i.updateMostRecentValidatorIP(nodeID, ip)
+	i.addGossipableIP(nodeID, ip)
+}
+
+func (*ipTracker) OnValidatorWeightChanged(ids.NodeID, uint64, uint64) {}
+
+func (i *ipTracker) OnValidatorRemoved(nodeID ids.NodeID, _ uint64) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	if i.manuallyTracked.Contains(nodeID) {
+		return
+	}
+
+	delete(i.mostRecentValidatorIPs, nodeID)
+	i.validators.Remove(nodeID)
+	i.removeGossipableIP(nodeID)
+}
+
+func (i *ipTracker) updateMostRecentValidatorIP(nodeID ids.NodeID, ip *ips.ClaimedIPPort) {
+	i.mostRecentValidatorIPs[nodeID] = ip
+	oldCount := i.bloomAdditions[nodeID]
+	if oldCount >= maxIPEntriesPerValidator {
+		return
+	}
+
+	// If the validator set is growing rapidly, we should increase the size of
+	// the bloom filter.
+	if count := i.bloom.Count(); count >= i.maxBloomCount {
+		if err := i.resetBloom(); err != nil {
+			i.log.Error("failed to reset validator tracker bloom filter",
+				zap.Int("maxCount", i.maxBloomCount),
+				zap.Int("currentCount", count),
+				zap.Error(err),
+			)
+		} else {
+			i.log.Info("reset validator tracker bloom filter",
+				zap.Int("currentCount", count),
+			)
+		}
+		return
+	}
+
+	i.bloomAdditions[nodeID] = oldCount + 1
+	gossipID := ip.GossipID()
+	bloom.Add(i.bloom, gossipID[:], i.bloomSalt[:])
+}
+
+func (i *ipTracker) addGossipableIP(nodeID ids.NodeID, ip *ips.ClaimedIPPort) {
+	i.gossipableIndicies[nodeID] = len(i.gossipableIPs)
+	i.gossipableIPs = append(i.gossipableIPs, ip)
+}
+
+func (i *ipTracker) removeGossipableIP(nodeID ids.NodeID) {
+	indexToRemove, wasGossipable := i.gossipableIndicies[nodeID]
+	if !wasGossipable {
+		return
+	}
+
+	newNumGossipable := len(i.gossipableIPs) - 1
+	if newNumGossipable != indexToRemove {
+		replacementIP := i.gossipableIPs[newNumGossipable]
+		replacementNodeID := replacementIP.NodeID()
+		i.gossipableIndicies[replacementNodeID] = indexToRemove
+		i.gossipableIPs[indexToRemove] = replacementIP
+	}
+
+	delete(i.gossipableIndicies, nodeID)
+	i.gossipableIPs[newNumGossipable] = nil
+	i.gossipableIPs = i.gossipableIPs[:newNumGossipable]
+}
+
+func (i *ipTracker) GetValidatorIPs(exceptNodeID ids.NodeID, exceptIPs *bloom.ReadFilter, salt ids.ID, maxNumIPs int) []*ips.ClaimedIPPort {
+	var (
+		uniform = sampler.NewUniform()
+		ips     = make([]*ips.ClaimedIPPort, 0, maxNumIPs)
+	)
+
+	i.lock.RLock()
+	defer i.lock.RUnlock()
+
+	uniform.Initialize(uint64(len(i.gossipableIPs)))
+	for len(ips) < maxNumIPs {
```

Comment:
May be good to callout that this can lead to us iterating over all the IPs we know about. I assume the bloom filter can run through this pretty fast.